File: src/pratice/graph/bfs/토마토7569.java
package pratice.graph.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class 토마토7569 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int H, N, M;
    public static int[][][] box;
    public static int[][][] distances;
    public static int[][] dir ={
            {0, 1, 0}, //  상
            {0, -1, 0}, // 하
            {0, 0, 1}, //우
            {0, 0, -1},  // 좌
            {1, 0, 0}, // 위
            {-1, 0, 0} // 아래
    };
    public static Queue<Point> needVisited = new LinkedList<>();

    public static void main(String[] args) throws IOException {
        input();
        pro();

    }

    private static void pro() {
        while(!needVisited.isEmpty()){
            Point p = needVisited.poll();
            for (int i = 0; i < 6; i++) {
                int x = p.x + dir[i][0];
                int y = p.y + dir[i][1];
                int z = p.z + dir[i][2];
                if(x >= 0 && y >= 0 && x < H && y < N && z >= 0 && z < M){
                    if(distances[x][y][z] == 0 && box[x][y][z] == 0){
                        needVisited.add(new Point(x, y, z));
                        distances[x][y][z] = distances[p.x][p.y][p.z] + 1;
                        box[x][y][z] = 1;
                    }
                }
            }
        }
        System.out.println(fun());
    }

    private static int fun() {
        int max = -1;
        for (int i = 0; i < H; i++) {
            for (int j = 0; j < N; j++) {
                for (int k = 0; k < M; k++) {
                    if(box[i][j][k] == 0){
                        return - 1;
                    }
                    max =  Math.max(distances[i][j][k], max);
                }
            }
        }
        return max;
    }

    private static void input() throws IOException {
        String[] split = br.readLine().split(" ");
        M = Integer.parseInt(split[0]);
        N = Integer.parseInt(split[1]);
        H = Integer.parseInt(split[2]);
        box = new int[H][N][M];
        distances = new int[H][N][M];
        for (int i = 0; i < H; i++) {
            for (int j = 0; j < N; j++) {
                split = br.readLine().split(" ");
                for (int k = 0; k < M; k++) {
                    box[i][j][k] = Integer.parseInt(split[k]);
                    if(box[i][j][k] == 1){
                        needVisited.add(new Point(i, j, k));
                    }
                }
            }
        }
    }
    static class Point{
        int x;
        int y;
        int z;

        public Point(int x, int y, int z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }
    }
}

File: src/pratice/graph/bfs/토마토7569.kt
package pratice.graph.bfs

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

private var br = BufferedReader(InputStreamReader(System.`in`))
private var H = 0
private var N = 0
private var M = 0
private lateinit var box: Array<Array<IntArray>>
private lateinit var distances: Array<Array<IntArray>>
private var dir = arrayOf(
    intArrayOf(0, 1, 0),
    intArrayOf(0, -1, 0),
    intArrayOf(0, 0, 1),
    intArrayOf(0, 0, -1),
    intArrayOf(1, 0, 0),
    intArrayOf(-1, 0, 0)
)
private var needVisited: Queue<Point> = LinkedList()
fun main(args: Array<String>) {
    input()
    pro()
}

private fun pro() {
    while (!needVisited.isEmpty()) {
        val p = needVisited.poll()
        for (i in 0..5) {
            val x = p.x + dir[i][0]
            val y = p.y + dir[i][1]
            val z = p.z + dir[i][2]
            if (x >= 0 && y >= 0 && x < H && y < N && z >= 0 && z < M) {
                if (distances[x][y][z] == 0 && box[x][y][z] == 0) {
                    needVisited.add(Point(x, y, z))
                    distances[x][y][z] = distances[p.x][p.y][p.z] + 1
                    box[x][y][z] = 1
                }
            }
        }
    }
    println(`fun`())
}

private fun `fun`(): Int {
    var max = -1
    for (i in 0 until H) {
        for (j in 0 until N) {
            for (k in 0 until M) {
                if (box[i][j][k] == 0) {
                    return -1
                }
                max = Math.max(distances[i][j][k], max)
            }
        }
    }
    return max
}

private fun input() {
    var split = br.readLine().split(" ".toRegex()).dropLastWhile { it.isEmpty() }
        .toTypedArray()
    M = split[0].toInt()
    N = split[1].toInt()
    H = split[2].toInt()
    box = Array(H) {
        Array(N) {
            IntArray(
                M
            )
        }
    }
    distances = Array(H) {
        Array(N) {
            IntArray(
                M
            )
        }
    }
    for (i in 0 until H) {
        for (j in 0 until N) {
            split = br.readLine().split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (k in 0 until M) {
                box[i][j][k] = split[k].toInt()
                if (box[i][j][k] == 1) {
                    needVisited.add(Point(i, j, k))
                }
            }
        }
    }
}

private class Point(var x: Int, var y: Int, var z: Int)
